### DISCUSSION TITLE
MeshCentral 3 wishlist

### DISCUSSION THREAD

**Discussion by rallisf1** [Ylianst/MeshCentral#3673] [Ideas]
I know MeshCentral 2 is still in beta (no v1) but it doesn't hurt planning what the next major release should be like. I'll go ahead with my suggestions; feel free to write yours below.

- Move most configuration options to database and make it possible to reload the configuration without MC restart (e.g. to add a new domain)
- Refactor the front-end with a modern framework like **Svelte**/Vue/React and TailwindCSS
- Branding through the gui (I know @JSuenram is already working on it and using something like TailwindCSS would make it a lot easier)
- Replace all icons with SVG alternatives
- Add external storage options for files: Google Drive, Dropbox, S3, FTP etc
- Create a single cross-platform agent (.net core via Xamarin can run pretty much anywhere, downside is the agent will become huge like 80+ MBs)
- Android & iOS remote control (currently we can only remote view android)

P.S. I can definetely help on the front-end stuff, I prefer svelte that's why it's bold but anything is better than current handlebars spaghetti code (sorry @Ylianst I know you are not a front-end guy and tbh I'm surprised you made it look this well as it is).

**Comment by Ylianst** [comment]
Actually, there was a MeshCentral v1 a long time back ([History video here](https://www.youtube.com/watch?v=4vME1owKGLw)). The server was coded in C# on ASP.NET using IIS and Microsoft SQL Server. I worked on that version for over 8 years, started before websocket even existed and had to do HTTP polling to emulate data streaming to web browsers. In any case, I am glad that code base is long gone for me. It was absolutely horrible to install, required Microsoft OS/IIS/DB. For a while I would the next version "MeshCentral v2", but now, it's just MeshCentral and v1 is a distant memory.

I 100% agree that the front end needs to be redone using a proper framework. No offense taking, a new revision of the front end would be high on my list too. I need to learn a framework and get on that at some point.

- Move most configuration options to database and make it possible to reload the configuration without MC restart (e.g. to add a new domain)

Even if it's not moved to the database, having a command to reload the config.json would be nice however, I am not sure everything can be done without a restart.

- Refactor the front-end with a modern framework like Svelte/Vue/React and TailwindCSS

Yes!

- Branding through the gui (I know @JSuenram is already working on it and using something like TailwindCSS would make it a lot easier)

Yes!

- Replace all icons with SVG alternatives

Yes! However, I am doing the best I can with low and high res images now, so high DPI displays will see nicer images.

- Add external storage options for files: Google Drive, Dropbox, S3, FTP etc

Interesting.

- Create a single cross-platform agent (.net core via Xamarin can run pretty much anywhere, downside is the agent will become huge like 80+ MBs)

This is probably not going to work as well as you think. Of course, C code can cross-compile widely and we compile to x86 & ARM in both 32 and 64 bit. We run DukTape on the agent so we have a high level language even in C code. It's difficult to come up with anything else that would be as compact and not require any dependencies on the host OS.

- Android & iOS remote control (currently we can only remote view android)

This is a limitation of Android. I have not worked on iOS because the limitation are even worst. The agent on Android runs as an app and is sandboxes by the OS. There could be ways around it, but not sure an open source project can do it.

**Reply by rallisf1** [reply]
Thanks for the thorough reply. I vouch for contributing in the new front end and might set up a bounty myself for the mobile apps. Reloading domains without restart would be great!

**Comment by JSuenram** [comment]
Regarding to the "spaghetti code", yes it is really hard at some points if you start contributing to MeshCentral.
I had also Pushed some bugs, which thanks to @Ylianst get caught, before I messed it up totally.

But I get some basic knowledge about the does and don't the last days. 

What I can say.... it does not look perfect, but holy shit... the code is blazing FAST!

**Comment by HuntTheSun** [comment]
My only feature request would be a alternative way of sending keyboard shortcuts that are not easily transferred through the browser UI, 
maybe adding a keyboard shortcut to send keystrokes like Win+R, maybe through ALT+Another Key(s).
It is only a minor gripe of mine (maybe there is a similar convenient feature for this and I just havent found it yet), so to me it is just a minor issue of convenience.

I cant say enough how much I love this software, really great work.

**Comment by nccerhostmaster** [comment]
A pretty front end would be nice but PLEASE don't create a slower, less efficient UI in pursuit of some swoopy design paradigm. Seen this happen too many times.
